File: WarmUp_Days/warmup-day-1-basics/reverse-a-number/solutions/reverse_a_number.cpp
// Problem: Reverse a Number
// Difficulty: Easy


#include <iostream>
#include <climits>
using namespace std;

long long reverse_number(long long n) {
    long long sign = n < 0 ? -1 : 1;
    n = llabs(n);
    
    long long rev = 0;
    while (n) {
        rev = rev * 10 + (n % 10);
        n /= 10;

        // Overflow detection before multiplying or adding
        if (rev > (LLONG_MAX - n % 10) / 10) {
            cerr << "Overflow detected!\n";
            return 0;
        }
    }
    return sign * rev;
}

int main() {
    cout << reverse_number(1234) << "\n";
    return 0;
}

File: WarmUp_Days/warmup-day-1-basics/reverse-a-number/solutions/reverse_a_number.py
"""
Problem: Reverse a Number
Difficulty: Easy

"""

def reverse_number(n: int) -> int:
    # TODO: implement correctly handling signs and overflow if needed
    INT_MAX = 2**31 - 1
    INT_MIN = -2**31

    sign = -1 if n < 0 else 1
    n = abs(n)
    rev = 0
    while n:
        rev = rev * 10 + n % 10
        n //= 10

        # Overflow detection
        if rev > (INT_MAX - (n % 10)) // 10:
            print("Overflow detected!")
            return 0
        

    return sign * rev

    # Clamp again in case of signed overflow
    if rev < INT_MIN or rev > INT_MAX:
        print("Overflow detected!")
        return 0

if __name__ == "__main__":
    print(reverse_number(12345))
